### PR TITLE
Add an option validation on cli commands

### PR DESCRIPTION
**Pull request by aledeg**
If an option used on cli is not recognized, the command
aborts and displays an error message.
If the typed option is similar to one of the recognized
options, a hint is displayed.

At the moment, there is a limitation on long options.
Short options are not validated at the moment.

See #2046

### CODE REVIEW

File: cli/_cli.php [hunk 1]
```diff
@@ -64,3 +67,30 @@ function performRequirementCheck($databaseType) {
...
+
+	if (0 !== count($unknownOptions)) {
+	    $unknownOption = array_shift($unknownOptions);
+	    $possibleOption = array_filter($sanitizeParams, function($a) use ($unknownOption) {
+	        return levenshtein($a, $unknownOption) < 2;
+	    });
+	    if (0 === count($possibleOption)) {
+            fail(sprintf('FreshRSS error: unknown option: %s'), $unknownOption);
+        } else {
+            fail(sprintf('FreshRSS error: unknown option: %s. Do you mean %s?', $unknownOption, array_shift($possibleOption)));
```
Comment (Frenzie) [inline cli/_cli.php]: I'd say "Did you mean" or perhaps "The most similar command/option/flag is:".

Side note, what if there are multiple results with a low Levenshtein distance?

**Review by Frenzie — COMMENTED** [review]
Interesting idea, although given the fairly limited number of options I'm not entirely convinced that it's actually more helpful than the usage notice. :-)

File: cli/_cli.php (hunk 1 above)
Comment (aledeg) [reply cli/_cli.php]: Noted.
In the case of multiple results, the first one is displayed

**Comment by aledeg** [comment]
I can remove the the suggestion part. It was just an experiment.
But the validation part is interesting because if one of the options is inexistent, the following one are discarded. In that case, you could end up doing things that you don't want to do.
See related issue for example.

**Comment by Frenzie** [comment]
You could also consider printing both, so you'd get something like this:

```
Did you mean --bla

Usage notice: --bla --yada
```

**Comment by aledeg** [comment]
Considering how the cli is build, it won't be possible to have both

**Comment by Frenzie** [comment]
I think usage info is more useful personally. Any thoughts @Alkarex ?

**Comment by aledeg** [comment]
I can do that but that means a lot of changes

**Comment by Frenzie** [comment]
I don't think suggestion is very useful though. :-)

(At least not when there are fewer than 10 options.)

**Comment by aledeg** [comment]
@Frenzie I've changed the behavior to display wrong parameters AND the usage. Let me know what you think!

File: cli/_cli.php [hunk 2]
```diff
@@ -64,3 +67,25 @@ function performRequirementCheck($databaseType) {
...
+function validateOptions($input, $params) {
+	$sanitizeInput = getLongOptions($input, REGEX_INPUT_OPTIONS);
+	$sanitizeParams = getLongOptions($params, REGEX_PARAM_OPTIONS);
+	$unknownOptions = array_diff($sanitizeInput, $sanitizeParams);
+
+	if (0 === count($unknownOptions)) {
+		return true;
+	}
+
+	fwrite(STDERR, sprintf("FreshRSS error: unknown options: %s\n", implode (', ', $unknownOptions)));
```
Comment (Frenzie) [inline cli/_cli.php:89]: I suppose you could consider giving `fail()` an argument regarding whether to die or not (true by default), or create a non-`die()`ing duplicate like a `warning()` function. But that might be premature abstraction.

**Review by Frenzie — APPROVED** [review]
lgtm!

File: cli/_cli.php (hunk 2 above)
Comment (aledeg) [reply cli/_cli.php:89]: That's a good point. Maybe we should do that if we intend to add more warnings.

**Comment by Alkarex** [comment]
@aledeg Could you please fix the Travis white-space error? :-)

**Comment by aledeg** [comment]
@Alkarex. Will do!

**Comment by aledeg** [comment]
@Alkarex it's all good!

**Comment by oupala** [comment]
As far as I have tested this merge request, it fixes the bug mentioned in #2046.

Thanks @aledeg for fixing this.

I still have 2 remarks:

- it is very strange to get syntax inconsistency in parameter name: `--api_password` seems not natural as we usually use dash *or* underscore, but not both of them, we should convert all parameter to dash syntax: `--api-password`
- it is also very strange to see that installing FreshRSS on an already installed FreshRSS returns an error, following idempotency we should more return a `0` return code:

```bash
>php cli/do-install.php --default_user myuser --language fr --api-enabled --db-type sqlite
FreshRSS looks to be already installed! Please use `./cli/reconfigure.php` instead.
>echo $?
1
```

**Pull request merged**